src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java

```diff
@@ -483,6 +435,32 @@ public void testGetShardCacheAgeEmptyForcesUpdate() {
         verify(mockClient).listShards(any());
     }
 
+    @Test
+    public void testNoDuplicateShardsInPartialFailure() {
+        proxy.setCachedShardMap(null);
+        ListShardsResult firstPage = new ListShardsResult().withShards(shards.subList(0, 2)).withNextToken(NEXT_TOKEN);
+        ListShardsResult lastPage = new ListShardsResult().withShards(shards.subList(2, shards.size())).withNextToken(null);
+        when(mockClient.listShards(any())).thenReturn(firstPage).thenThrow(new RuntimeException("Failed!")) // First call
+                .thenReturn(firstPage).thenReturn(lastPage); // second call
+
+        try {
+            proxy.getShardList();
+        } catch (Exception e) {
+            // Do nothing
+        }
+        assertEquals(shards, proxy.getShardList());
+    }
+
+    @Test
+    public void testDuplicateShardResponseDedupedGracefully() {
+        proxy.setCachedShardMap(null);
+        ListShardsResult pageOfShards = new ListShardsResult().withShards(shards).withNextToken(null);
```

Comment:
Does the "shards" list include duplicates? I don't think we are testing the duplicate case here.

Reply:
Your'e right... duplicated the shard in next revision...